[FEATURE] Add Semi-SL multilabel classification algorithm (#1805)
open-edge-platform/training_extensions: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst

| EfficientNet-V2-S     | 91.91           | 77.28     | 71.52            | 80.24     |
+-----------------------+-----------------+-----------+------------------+-----------+

************************
Semi-supervised Learning
************************

Semi-SL (Semi-supervised Learning) is a type of machine learning algorithm that uses both labeled and unlabeled data to improve the performance of the model. This is particularly useful when labeled data is limited, expensive or time-consuming to obtain.

> To utilize unlabeled data during training, we use `BarlowTwins loss <https://arxiv.org/abs/2103.03230>`_ as an auxiliary loss for Semi-SL task solving. BarlowTwins enforces consistency across augmented versions of the same data (both labeled and unlabeled). Additionally, our algorithm uses a combination of strong data augmentations and a specific optimizer called Sharpness-Aware Minimization (SAM) to further improve the accuracy of the model.

Reviewer: ```suggestion
To utilize unlabeled data during training multilabel classification model, we use `BarlowTwins loss <https://arxiv.org/abs/2103.03230>`_ as an auxiliary loss for Semi-SL  task solving. BarlowTwins enforces consistency across augmented versions of the same data (both labeled and unlabeled). Additionally, our algorithm uses a combination of strong data augmentations and a specific optimizer called Sharpness-Aware Minimization (SAM) to further improve the accuracy of the model.
```

Reviewer: Please add link for SAM

> Overall, OpenVINO™ Training Extensions utilizes powerful techniques for improving the performance of Semi-SL algorithm with limited labeled data. They can be particularly useful in domains where labeled data is expensive or difficult to obtain, and can help to reduce the time and cost associated with collecting labeled data.

Reviewer: I think this sentence is the same as the first one and doesn't give  us additional info. Consider to remove it

Author: Deleted

.. _mlc_cls_semi_supervised_pipeline:

- ``BarlowTwins loss``: A specific implementation of Semi-SL that combines the use of a consistency loss with strong data augmentations, and a specific optimizer called Sharpness-Aware Minimization (SAM) to improve the performance of the model.

- ``Adaptive loss auxiliary loss weighting``: A technique for assigning such a weight for an auxiliary loss that the resulting value is a predefined fraction of the EMA-smoothed main loss value. This method allows aligning contribution of the losses during different training phases.

- ``Exponential Moving Average (EMA)``: A technique for maintaining a moving average of the model's parameters, which can improve the generalization performance of the model.

> - ``Additional techniques``: Other than that, we use several solutions that apply to supervised learning (No bias Decay, Augmentations, Early-Stopping, etc.)

Reviewer: You also use "strong" and "weak" pipelines, could you describe them here?

Author: I described them in general description to clarify work of the BT loss.

Please, refer to the :doc:`tutorial <../../../tutorials/advanced/semi_sl>` on how to train semi-supervised learning.
Training time depends on the number of images and can be up to several times longer than conventional supervised learning.

> In the table below the mAP on some academic datasets using our pipeline is presented.

Reviewer: ```suggestion
In the table below the mAP on some academic datasets using our pipeline is presented.
```
```suggestion
In the table below the mAP metric on some public datasets using our pipeline is presented.
```
I am not sure that Aerial Maritime is popular academic benchmark dataset

Reviewer: Also, please, describe more in depth how many images you utilized for this experiments

|        Dataset        | AerialMaritime 3 cls |         | VOC 2007 3 cls |         | COCO 14 3 cls  |         |
+=======================+======================+=========+================+=========+================+=========+
|                       |   SL                 | Semi-SL |  SL            | Semi-SL |   SL           | Semi-SL |
+-----------------------+----------------------+---------+----------------+---------+----------------+---------+
| MobileNet-V3-large-1x |  74.28               |  74.41  | 96.34          |  97.29  |  82.39         |  83.77  |
+-----------------------+----------------------+---------+----------------+---------+----------------+---------+
|   EfficientNet-B0     |  79.59               |  80.91  | 97.75          |  98.59  | 83.24          |  84.19  |
+-----------------------+----------------------+---------+----------------+---------+----------------+---------+
|  EfficientNet-V2-S    |  75.91               |  81.91  | 95.65          |  96.43  | 85.19          |  84.24  |
+-----------------------+----------------------+---------+----------------+---------+----------------+---------+

> AerialMaritime was sampled with 5 images per label. VOC and COCO were sampled with 40 images per label.

Reviewer: ```suggestion
AerialMaritime was sampled with 5 images per category. VOC and COCO were sampled with 40 images per category.
```
or class

.. note::

>     This result can vary greatly depending on the image selected for each class. Also, since there are few labeled settings for the Semi-SL algorithm. Some models may require larger datasets for better results.

Reviewer: ```suggestion
    This result greatly depends on the image selected for each class. Also, since there are few labeled settings for the Semi-SL algorithm. Some models may require larger datasets for better results.
```
